### PR TITLE
feature(events): add to event nemesis sub-context

### PR DESCRIPTION
**Pull request by juliayakovlev**
If an error event happens during nemesis, attach the nemesis information to the event.

Event message example:
```
2022-09-14 08:44:14.893: (FullScanEvent Severity.ERROR) period_type=end event_id=d23078b5-732c-4649-a909-25144eb22af8 during_nemesis=RestartWithResharding duration=7m27s node=longevity-10gb-3h-test-add-db-node-8ca92eee-2 select_from=keyspace1.standard1 message=FullScanThread operation ended successfully
```

in raw event:
```
{"base": "FullScanEvent", "type": null, "subtype": null, "ks_cf": "keyspace1.standard1", "message": "FullScanThread operation ended successfully", "node": "longevity-10gb-3h-test-add-db-node-8ca92eee-2", "shard": null, "event_timestamp": 1663145054.893798, "source_timestamp": null, "severity": "NORMAL", "event_id": "d23078b5-732c-4649-a909-25144eb22af8", "log_level": 20, "subcontext": [{"event_id": "f3ac8255-5df4-401e-935d-1d1fccc3300a", "base": "DisruptionEvent", "nemesis_name": "RestartWithResharding", "node": "Node longevity-10gb-3h-test-add-db-node-8ca92eee-2 [13.48.84.78 | 10.0.3.238] (seed: False)", "timestamp": 1663144652.646285}], "errors": [], "publish_event": true, "begin_timestamp": 1663144606.9619968, "end_timestamp": 1663145054.893798, "is_skipped": false, "skip_reason": "", "period_type": "end"}
```

Refs: https://github.com/scylladb/qa-tasks/issues/341

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Comment by fruch** [comment]
@juliayakovlev do you have a small run with it ?, so we can see the output from some real events looks like with it ?

File: sdcm/sct_events/base.py [hunk 1]
```diff
@@ -59,6 +60,19 @@ def __set_name__(self, owner: Type[SctEvent], name: str) -> None:
         self[owner.__name__] = owner  # add owner class to the registry.
 
 
+@dataclass
+class SubContext:
```
Comment (fruch) [inline sdcm/sct_events/base.py:64]: name is a bit too general, maybe `SubContextEvent` ?

File: sdcm/sct_events/continuous_event.py [hunk 2]
```diff
@@ -67,6 +67,19 @@ def del_event(self, event: ContinuousEvent):
     def find_continuous_events_by_hash(self, continuous_hash: int) -> list[ContinuousEvent]:
         return self.hashed_continuous_events.get(continuous_hash, []).copy()
 
+    def find_running_disruption_events(self):
+        all_nemeses = [(event[0].begin_timestamp, event[0].continuous_hash_tuple[1], event[0].nemesis_name,
```
Comment (fruch) [inline sdcm/sct_events/continuous_event.py:71]: why first convert the events to tuples, and then convert it to a dict ?
why not do it in one pass ?

File: sdcm/sct_events/continuous_event.py [hunk 3]
```diff
@@ -67,6 +67,19 @@ def del_event(self, event: ContinuousEvent):
...
+                        event[0].node, event[0].base, event[0].timestamp)
+                       for event in self.hashed_continuous_events.values()
+                       if event and event[0].base == "DisruptionEvent"]
+        if not all_nemeses:
+            return None
+
+        last_nemesis_event = [{'event_id': nemesis_event[1], 'nemesis_name': nemesis_event[2],
```
Comment (fruch) [inline sdcm/sct_events/continuous_event.py:78]: variable is called `last_nemesis_event` but it's a list, not one event.

File: sdcm/sct_events/base.py [hunk 4]
```diff
@@ -190,6 +205,36 @@ def timestamp(self) -> float:
...
+
+        during_nemesis = []
+        for event_context in self.subcontext:
+            if event_context["base"] != "DisruptionEvent":
+                continue
+
+            during_nemesis.append(event_context['nemesis_name'])
+
+        if during_nemesis:
+            fmt += f" during_nemesis={','.join(during_nemesis)}"
```
Comment (fruch) [inline sdcm/sct_events/base.py:220]: why not call it `context=` ? 

and put the creating of the context in it's own function  `contextfmt` ? where you can extend it and make it handle more types of nemesis ?

File: sdcm/sct_events/base.py [hunk 5]
```diff
@@ -59,6 +60,19 @@ def __set_name__(self, owner: Type[SctEvent], name: str) -> None:
...
 
 
+@dataclass
+class SubContext:
+    event_id: str
+    base: str
+    node: str
+    timestamp: float
+    nemesis_name: str
+    nodetool_command: str
```
Comment (fruch) [inline sdcm/sct_events/base.py:70]: why `nodetool_command` is here ?, doesn't seems to be used ?

File: sdcm/sct_events/base.py [hunk 6]
```diff
@@ -59,6 +60,19 @@ def __set_name__(self, owner: Type[SctEvent], name: str) -> None:
         self[owner.__name__] = owner  # add owner class to the registry.
 
 
+@dataclass
+class SubContext:
+    event_id: str
+    base: str
+    node: str
+    timestamp: float
+    nemesis_name: str
```
Comment (fruch) [inline sdcm/sct_events/base.py:69]: this is very specific to one type of Event, maybe there should be a dict of extra parameters ?

or add a `dict` function to the BaseEvent ? and remove the `SubContext` ?

File: sdcm/sct_events/continuous_event.py (hunk 2 above)
Comment (juliayakovlev) [reply sdcm/sct_events/continuous_event.py]: changed

File: sdcm/sct_events/base.py (hunk 6 above)
Comment (juliayakovlev) [reply sdcm/sct_events/base.py]: removed

File: sdcm/sct_events/base.py [hunk 7]
```diff
@@ -231,6 +268,10 @@ def dont_publish(self):
     def ready_to_publish(self):
         self._ready_to_publish = True
 
+    @cached_property
+    def event_to_json(self):
```
Comment (fruch) [inline sdcm/sct_events/base.py:272]: the name of this is very confusing, i.e. it's not clear how it's different from `to_json`,

maybe have a `subcontext_field` that you could use to filter the output of `to_json` of an event.

File: sdcm/sct_events/base.py [hunk 8]
```diff
@@ -190,6 +191,42 @@ def timestamp(self) -> float:
...
+    @property
+    def msgfmt(self):
+        fmt = "({0.base} {0.severity}) period_type={0.period_type} event_id={0.event_id}"
+
+        if during_nemesis := self.concatenate_subcontext_for_message(context_name='nemesis_name'):
+            fmt += f" during_nemesis={during_nemesis}"
+
+        return fmt
+
+    def add_subcontext(self, registry):
```
Comment (fruch) [inline sdcm/sct_events/base.py:216]: if on all calls to this function we are using `ContinuousEventsRegistry()` which is a singleton, why we need this as a parameter here ? do we assume that there would be multiple of those registries ?

and even if so, I would wait until we have that use case, and see how it's fit, and only then extend those functions for it.

**Comment by fruch** [comment]
@enaydanov can you give this one a look ?

File: sdcm/sct_events/base.py (hunk 8 above)
Comment (juliayakovlev) [reply sdcm/sct_events/base.py]: changed

File: sdcm/sct_events/base.py (hunk 7 above)
Comment (juliayakovlev) [reply sdcm/sct_events/base.py:274]: changed

File: sdcm/sct_events/base.py (hunk 8 above)
Comment (juliayakovlev) [reply sdcm/sct_events/base.py]: changed

**Comment by juliayakovlev** [comment]
Test passed https://jenkins.scylladb.com/job/scylla-staging/job/yulia/job/staging-10gb-3h/332/

**Review by fruch — APPROVED** [review]
LGTM

**Pull request merged**